pkg/operator/machinedeletionhooks/machinedeletionhooks.go

```diff
@@ -0,0 +1,293 @@
+	if err != nil {
+		return err
+	}
+	masterMachines, err := c.masterMachineLister.List(c.masterMachineSelector)
+	if err != nil {
+		return err
+	}
+	// machines with the deletion hooks are the ones that host etcd members
+	memberMachines := filterMachinesWithMachineDeletionHook(masterMachines)
+
+	// if the number of master machines is > controlPlane.replicas based on data from the cache
+	if len(memberMachines) <= desiredControlPlaneReplicasCount {
+		klog.V(4).Infof("haven't found a replacement machine, the number of desired control plane replicas: %d must be greater than the current number of machines that host an etcd member: %d", desiredControlPlaneReplicasCount, len(memberMachines))
+		return nil
+	}
+
+	// if at least one member machine is pending deletion
+	memberMachinesPendingDeletion := filterMachinesPendingDeletion(memberMachines)
+	if len(memberMachinesPendingDeletion) == 0 {
+		klog.V(4).Infof("currently, we don't have a machine pending deletion")
+		return nil
+	}
+
+	// get the members, issue a live call instead of reading from the config map
+	// because we need to take into account learners
+	members, err := c.etcdClient.MemberList(ctx)
+	if err != nil {
+		return err
+	}
+	currentMemberIPListSet := sets.NewString()
+	for _, member := range members {
+		votingMemberIP, err := ceohelpers.MemberToNodeInternalIP(member)
+		if err != nil {
+			return err
+		}
+		currentMemberIPListSet.Insert(votingMemberIP)
+	}
+
+	// find a machine that doesn't have a member
+	for memberMachinePendingDeletionIP, memberMachinePendingDeletion := range indexMachinesByNodeInternalIP(memberMachinesPendingDeletion) {
+		if currentMemberIPListSet.Has(memberMachinePendingDeletionIP) {
+			// Ignore machines that still have a member present
+			continue
+		}
+		memberMachinePendingDeletionCopy := memberMachinePendingDeletion.DeepCopy()
+		removeMachineDeletionHook(memberMachinePendingDeletionCopy)
+		if _, err := c.machineClient.Update(ctx, memberMachinePendingDeletionCopy, metav1.UpdateOptions{}); err == nil {
+			klog.V(2).Infof("successfully removed the deletion hook from machine %v", memberMachinePendingDeletion.Name)
+			recorder.Eventf("MachineDeletionHook", "successfully removed the deletion hook from machine %v as it doesn't have a member", memberMachinePendingDeletion.Name)
+		}
+		return err
+	}
+	return nil
+}
+
+// addDeletionHookToMasterMachines adds the Machine Deletion Hook to master machines
+func (c *machineDeletionHooksController) addDeletionHookToMasterMachines(ctx context.Context) error {
+	masterMachines, err := c.masterMachineLister.List(c.masterMachineSelector)
+	if err != nil {
+		return err
+	}
+	for _, masterMachine := range masterMachines {
+		if masterMachine.DeletionTimestamp != nil {
+			continue
+		}
+		if hasCompleteMachineDeletionHook(masterMachine) {
+			continue
+		}
+		votingMemberMachineCopy := masterMachine.DeepCopy()
```

Comment:
this is just `machineMemberCopy`, now right?

Reply:
yes, will update